From c9b2c4f9daf3a97589b754ac96d2168b244e004c Mon Sep 17 00:00:00 2001
From: Hanna Hawa <hannah@marvell.com>
Date: Sun, 8 Mar 2015 19:40:03 +0200
Subject: [PATCH 0140/1240] spl: 38x: added support for SPL for A38x SoC

	- Added support for SPL for A38x SoC
	- Change base address for SPL to 0xd0000000
	- Set the TEXT base of the SPL to 0x40000040

Change-Id: I7bc7101d884bab3df0228db3bd02f341846b09ce
Signed-off-by: Hanna Hawa <hannah@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/17314
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
---
 arch/arm/Kconfig                                 |  1 +
 arch/arm/include/asm/arch-armada38x/memory-map.h |  5 ++++-
 configs/mvebu_armada38x_defconfig                |  9 +++++----
 include/configs/armada38x.h                      | 13 +++++++++++++
 include/configs/mvebu-common.h                   |  7 +++++++
 scripts/Makefile.spl                             |  2 +-
 6 files changed, 31 insertions(+), 6 deletions(-)

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index 60ef9c8..49ba49a 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -107,6 +107,7 @@ config TARGET_ARMADA_8K
 
 config TARGET_ARMADA_38X
 	bool "Support Armada 38x"
+	select SUPPORT_SPL
 
 config TARGET_ARMADA_LP
 	bool "Support Armada LP"
diff --git a/arch/arm/include/asm/arch-armada38x/memory-map.h b/arch/arm/include/asm/arch-armada38x/memory-map.h
index 0674e23..7a560ff 100644
--- a/arch/arm/include/asm/arch-armada38x/memory-map.h
+++ b/arch/arm/include/asm/arch-armada38x/memory-map.h
@@ -26,8 +26,11 @@
 /* 0xF1000000 - 0xF1100000 - INTERNAL_REG */
 /* 0xFD000000 - 0xFD100000 - NOR_CS */
 
-/* General */
+#ifndef CONFIG_SPL_BUILD
 #define MVEBU_REGS_BASE			(0xF1000000)
+#else
+#define MVEBU_REGS_BASE                 (0xD0000000)
+#endif
 #define MVEBU_REGS_SIZE			(0x100000)
 
 /* Memory windows settings */
diff --git a/configs/mvebu_armada38x_defconfig b/configs/mvebu_armada38x_defconfig
index 65ee816..551f103 100644
--- a/configs/mvebu_armada38x_defconfig
+++ b/configs/mvebu_armada38x_defconfig
@@ -1,7 +1,8 @@
-CONFIG_ARM=y
-CONFIG_TARGET_ARMADA_38X=y
-CONFIG_MVEBU_UART=y
-CONFIG_MVEBU_I2C=y
+CONFIG_SPL=y
++S:CONFIG_ARM=y
++S:CONFIG_TARGET_ARMADA_38X=y
++S:CONFIG_MVEBU_UART=y
++S:CONFIG_MVEBU_I2C=y
 CONFIG_OF_CONTROL=y
 CONFIG_OF_EMBED=y
 CONFIG_DEFAULT_DEVICE_TREE="armada-385-db"
diff --git a/include/configs/armada38x.h b/include/configs/armada38x.h
index d150d7f..aa38003 100644
--- a/include/configs/armada38x.h
+++ b/include/configs/armada38x.h
@@ -101,6 +101,19 @@
 #define MV_DRAM_AUTO_SIZE
 #define CONFIG_SYS_DRAM_BANKS           1
 
+/* SPL */
+/* Defines for SPL */
+#define CONFIG_SPL_TEXT_BASE		0x40000040
+#define CONFIG_SPL_MAX_SIZE		(0x1ffc0)
+
+#define CONFIG_SPL_BSS_START_ADDR	(0x40020000)
+#define CONFIG_SPL_BSS_MAX_SIZE		(0x8000)
+
+#define CONFIG_SYS_SPL_MALLOC_START	(CONFIG_SPL_BSS_START_ADDR + \
+					 CONFIG_SPL_BSS_MAX_SIZE)
+#define CONFIG_SYS_SPL_MALLOC_SIZE	(0x4000)
+#define CONFIG_SPL_LDSCRIPT		"arch/arm/cpu/mvebu-common/u-boot-armv7-spl.lds"
+
 /* Include AFTER since it is affected by defines above */
 #include "mvebu-common.h"
 
diff --git a/include/configs/mvebu-common.h b/include/configs/mvebu-common.h
index d08dafc..10afaed 100644
--- a/include/configs/mvebu-common.h
+++ b/include/configs/mvebu-common.h
@@ -67,6 +67,13 @@
 #define CONFIG_ENV_OVERWRITE            /* allow to change env parameters */
 #undef  CONFIG_WATCHDOG                 /* watchdog disabled */
 
+/* SPL */
+/* Common defines for SPL */
+#define CONFIG_SPL_FRAMEWORK
+#define CONFIG_SPL_LIBCOMMON_SUPPORT
+#define CONFIG_SPL_LIBGENERIC_SUPPORT
+#define CONFIG_SPL_SERIAL_SUPPORT
+
 /* Cache */
 #define CONFIG_SYS_DCACHE_OFF
 #define CONFIG_SYS_CACHELINE_SIZE       32
diff --git a/scripts/Makefile.spl b/scripts/Makefile.spl
index 4c5bc7f..1b70bb8 100644
--- a/scripts/Makefile.spl
+++ b/scripts/Makefile.spl
@@ -79,7 +79,7 @@ libs-$(CONFIG_SPL_WATCHDOG_SUPPORT) += drivers/watchdog/
 libs-$(CONFIG_SPL_USB_HOST_SUPPORT) += drivers/usb/host/
 libs-$(CONFIG_OMAP_USB_PHY) += drivers/usb/phy/
 libs-$(CONFIG_SPL_SATA_SUPPORT) += drivers/block/
-libs-y += drivers/ddr/
+libs-$(CONFIG_MVEBU_SPL_DDR_SUPPORT) += drivers/ddr/
 
 head-y		:= $(addprefix $(obj)/,$(head-y))
 libs-y		:= $(addprefix $(obj)/,$(libs-y))
-- 
1.9.1

